docs/standard/native-interop/interop-guidance.md

```diff
@@ -0,0 +1,247 @@
+   2. Copies the contents if `[In]` _(the default for a `StringBuilder` parameter)_  
+   3. Copies the native buffer into a newly allocated managed array if `[Out]` **{3}** _(also the default for `StringBuilder`)_  
+3. `ToString()` allocates yet another managed array **{4}**
+
+That is *{4}* allocations to get a string out of native code. The best you can do to limit this is to reuse the `StringBuilder`
+in another call but this still only saves *1* allocation. It is much better to use and cache a character buffer from `ArrayPool`- you can then get down to just the allocation for the `ToString()` on subsequent calls.
+
+The other issue with `StringBuilder` is that it always copies the return buffer back up to the first null. If the passed back string isn't terminated or is a double-null-terminated string your P/Invoke is incorrect at best.
+
+If you *do* use `StringBuilder` one last gotcha is that the capacity does **not** include a hidden null which is always accounted for in interop. It is pretty common for people to get this wrong as most APIs want the size of the buffer *including* the null. This can result in wasted/unnecessary allocations. Additionally, this gotcha prevents the runtime from optimizing `StringBuilder` marshalling to minimize copies.
+
+**✔️ CONSIDER** using `char[]`s from an `ArrayPool`.
+
+For more information on string marshalling, see the documentation on [Default Marshalling for Strings](../../framework/interop/default-marshaling-for-strings) and [customizing string marshalling](./customize-parameter-marshalling#customizing-string-parameters).
+
+> __Windows Specific__  
+> For `[Out]` strings the CLR will use `CoTaskMemFree` by default to free strings or `SysStringFree` for strings that are marked
+as `UnmanagedType.BSTR`.  
+> **For most APIs with an output string buffer:**  
+> The passed in character count must include the null. If the returned value is less than the passed in character count the call has succeeded and the value is the number of characters *without* the trailing null. Otherwise the count is the required size of the buffer *including* the null character.  
+> - Pass in 5, get 4: The string is 4 characters long with a trailing null.
+> - Pass in 5, get 6: The string is 5 characters long, need a 6 character buffer to hold the null.  
+> [Windows Data Types for Strings](/windows/desktop/Intl/windows-data-types-for-strings)
+
+## Boolean parameters and fields
+
+Booleans are easy to mess up. The default marshalling for .NET is as the Windows type `BOOL`, where it is a 4 byte value. However, the `_Bool`, and `bool` types in C and C++ are a *single* byte. This can lead to hard to track down bugs as half the return value will be discarded, which will only *potentially* change the result. See the documentation on [customizing boolean field marshalling](customize-struct-marshalling#customizing-boolean-field-marshalling) for information on marshalling .NET `bool` values to C or C++ `bool` types.
```

Comment:
```suggestion
Booleans are easy to mess up. The default marshalling for .NET is as the Windows type `BOOL`, where it's a 4-byte value. However, the `_Bool`, and `bool` types in C and C++ are a *single* byte. This can lead to hard to track down bugs as half the return value will be discarded, which will only *potentially* change the result. For more for information on marshalling .NET `bool` values to C or C++ `bool` types, see the documentation on [customizing boolean field marshalling](customize-struct-marshalling#customizing-boolean-field-marshalling).
```

Reply:
"The default marshalling for .NET is as the Windows type `BOOL`" is a bit hard to understand

Reply:
I'll clean up that wording